### DISCUSSION TITLE
PGP Decrypt Stream - NodeJs

### DISCUSSION THREAD

**Discussion by NicoBougrand** [openpgpjs/openpgpjs#1266] [Q&A]
Hello everyone,

Here is my issue.
I must retrieve a file, encrypted with pgp, from a ftp server, and put it decrypted on another ftp server.
For that, I would like to read the file from ftp, through a strem, decrypt it with openpgp, with a transformStream, and write it in the other ftp.

From FTP to FTP,  without the transformStream, all is ok.
When I add a transformStream which add something all is ok alose.

But when I try to decrypt the chunk in the transformStream, it doesn't work and I have the following issue : 

> _Error: Error during parsing. This message / key probably does not conform to a valid OpenPGP format_

I found no example of streaming decryption.
Has someone already face this issue to help me, or have any idea please.

Thanks

Here is my TransformStream

```
import fs from "fs";
import * as openpgp from "openpgp";
import { Stream } from "stream";

openpgp.config.debug = true;

export default class DecryptPGP extends Stream.Transform {

    public static async  instanciate(): Promise<DecryptPGP> {
        const publicKey = fs.readFileSync("./certs/pgp/public.asc", "utf8"); // encrypted private key
        const privateKey = fs.readFileSync("./certs/pgp/private.asc", "utf8"); // encrypted private key
        const passphrase = "PASSPHRASE"; // what the privKey is encrypted with

        const publicKeyObj = (await openpgp.key.readArmored(publicKey));
        const privKeyObj = (await openpgp.key.readArmored(privateKey));
        await privKeyObj.keys.forEach((key) => key.decrypt(passphrase));

        return new DecryptPGP(publicKeyObj, privKeyObj);
    }

    private publicKeyObj: openpgp.key.KeyResult;
    private privKeyObj: openpgp.key.KeyResult;
    private isArmored = false;
    private hasCheckedArmored = false;

    private constructor(publicKeyObj: openpgp.key.KeyResult, privKeyObj: openpgp.key.KeyResult) {
        super();
        this.publicKeyObj = publicKeyObj;
        this.privKeyObj = privKeyObj;
        openpgp.config.ignore_mdc_error = true;
    }

    public async _transform(chunk: any, encoding: string, cb: (error?: Error | null, data?: any) => void) {
        if (!this.hasCheckedArmored) {
            console.info("Chunk is armored");
            this.isArmored = typeof chunk === "string" && chunk.startsWith("-----BEGIN PGP MESSAGE-----");
            this.hasCheckedArmored = true;
        }

        const readableStream = new ReadableStream({
            start(controller) {
                controller.enqueue(chunk);
                controller.close();
            }
        });

        const options = {
            message: this.isArmored ? await openpgp.message.readArmored(readableStream) : await openpgp.message.read(readableStream),
            privateKeys: this.privKeyObj.keys
        };

        try {
            const decrypt = await openpgp.decrypt(options);
            // tslint:disable-next-line: no-null-keyword
            cb(null, chunk);
        }
        catch (e) {
            console.error("issue here", e);
            cb(e);
        }
    }
}

```

**Comment by twiss** [comment]
Hey @NicoBougrand :wave:
OpenPGP.js supports passing a Node stream:
```js
await openpgp.decrypt({
  message: await openpgp.message.readArmored(
    fs.createReadStream(path)
  )
});
```
That's probably the easiest way to stream-decrypt a file, but let me know if you need to use a TransformStream for some reason.

(Currently, you're trying to pass each chunk of the file to `openpgp.decrypt` individually, and making a new ReadableStream object for that each time. To make your code work, you have to pass a single ReadableStream object, and pass each chunk to its `controller.enqueue`.)

**Comment by NicoBougrand** [comment]
Hello @twiss ,

Thanks for the quick answer

I use a transform stream because, my global code is like this 

` const conn = new Client();
        const sshOpt = {
            /** Hostname or IP address of the server. */
            host: "sftphost",
            /** Username for authentication. */
            username: "user",
            /** Password for password-based user authentication. */
            password: "pwd*"
        };

        const remoteDir = "remoteDir";
        conn.on("ready", () => {
            conn.sftp(async (err, sftp) => {
                if (err) { throw err; }

                // const remoteFile = remoteDir + "/test-big-file.txt";
                const remoteFile = remoteDir + "/test-big-file.asc";

                const sftpFileStream = sftp.createReadStream(remoteFile);

                const decryptPGPFile = await DecryptPGP.instanciate();
                sftpFileStream.pipe(decryptPGPFile)// .pipe(process.stdout)
                    .on("end", () => {
                        Log.debug("c'est fini");
                    });

                const loadResult = await hdfsUtils.loadOnHDFSFolder(decryptPGPFile, "/file-loader/test-sftp/test-big-file-armored.txt", cancelSource);
                Log.debug("loadResult", loadResult);

            });
        });
        conn.on("error", (err) => {
            Log.error("SSH connection stream problem");
            throw err;
        });
        conn.connect(sshOpt);`

as you can see, I retrieve the file from sftp through a stream.
I must decrypt this stream, and give the loadOnHdfsFolder another stream...

It seems to me that decrypt function doesn't give back a stream (that I can pass to the loadOnHdfsFolder function). That's why I used a TransformStream.
Maybe it is not useful, but I have not seen any other way to make it work for now.

**Comment by twiss** [comment]
> It seems to me that decrypt function doesn't give back a stream

It returns an object with a `data` property, that should be a stream.

Please see https://github.com/openpgpjs/openpgpjs/releases/tag/v4.0.0 for some more details about streaming.

**Comment by NicoBougrand** [comment]
Thanks,
Effectively I have not seen this.

My main concern is now that I have type incompatibility  between my streams (that is not ok in  typescript)
Indeed 

- sftp.createReadStream return a stream.Readable
- openpgp.message.read take as input a ReadableStream<Unint8Array>
- openpgp.decrypt return a DecryptResult with data of type ReadableStream<Uint8Array>
- my function loadHDFS take a NodeJS.ReadableStream 

I try to map one type to the other, but no luck for now, that's why I wanted to create a new ReadableStream. but not sure it's the good solution...

**Comment by twiss** [comment]
> - openpgp.message.read take as input a ReadableStream
> - openpgp.decrypt return a DecryptResult with data of type ReadableStream

I know the documentation says so, but as mentioned this is not accurate, they can also take and return a Node stream.

So probably the proper solution to this is to fix the documentation and typescript definitions. See e.g. https://github.com/openpgpjs/openpgpjs/issues/561#issuecomment-530524005.

**Comment by NicoBougrand** [comment]
Yes fine, I didn't think about changing type to have no issue.
Thanks a lot
It works well now.

**Comment by lasergoat** [comment]
@twiss 

> Hey @NicoBougrand 👋
> OpenPGP.js supports passing a Node stream:
> 
> ```js
> await openpgp.decrypt({
>   message: await openpgp.message.readArmored(
>     fs.createReadStream(path)
>   )
> });
> ```
> 
> That's probably the easiest way to stream-decrypt a file, but let me know if you need to use a TransformStream for some reason.
> 
> (Currently, you're trying to pass each chunk of the file to `openpgp.decrypt` individually, and making a new ReadableStream object for that each time. To make your code work, you have to pass a single ReadableStream object, and pass each chunk to its `controller.enqueue`.)

For some reason, I was hung up on a completely different issue where I kept getting these errors based on changing the code and how I was treating the encrypted text (from a buffer, uint8array, plain text etc):

- This message / key probably does not conform to a valid OpenPGP format
- Session key decryption failed
- misformed armored text

But, somehow, treating the file as a read stream fixed the issue. Your comment saved me.

**Comment by shivanipremi** [comment]
> Hey @NicoBougrand 👋
> OpenPGP.js supports passing a Node stream:
> 
> ```js
> await openpgp.decrypt({
>   message: await openpgp.message.readArmored(
>     fs.createReadStream(path)
>   )
> });
> ```
> 
> That's probably the easiest way to stream-decrypt a file, but let me know if you need to use a TransformStream for some reason.
> 
> (Currently, you're trying to pass each chunk of the file to `openpgp.decrypt` individually, and making a new ReadableStream object for that each time. To make your code work, you have to pass a single ReadableStream object, and pass each chunk to its `controller.enqueue`.)

I am using this function, to decrypt a very large file (greater than 2 gb)

But, getting error while decrypting

   const readStream = fs.createReadStream(encryptedFilePath);

            let decryptedReadTransfromStreamObj = await openpgp.decrypt({
                message: await openpgp.message.read(readStream),
                privateKeys: [privateKeyObj],
                format: 'binary',
                streaming: 'node'
            });

Is there any other way I can decrypt a large file successfully?
I tried the above using transform stream, but not able to get correct decrypted file

**Comment by matthias-herrmann** [comment]
I got this errors above when using streams other than simple file streams. E.g. streaming from aws s3 to the decrypt function.

I resolved it by setting the right encoding and using `openpgp.readMessage`. This cost me several hours to figure it out. I'm happy that it works now.


```
const encryptedStream = s3Client.getObject({Bucket: 'bucket', Key: 'file.txt'}).createReadStream().setEncoding("utf-8");
    var message = await openpgp.readMessage({'armoredMessage': encryptedStream});
    
const {data} = await openpgp.decrypt({
     message: message,
     publicKeys: myPublicKey, // for verification (optional)
     privateKeys: otherPrivateKey, // for decryption,
     armor: true,
     streaming: "node",
});
```

**Comment by PujaR-dev** [comment]
I would suggest to go with 'gpg' package instead of 'opengpg':

`gpg.importKey(privateKey, [], (success, err) => {
                  gpg.decrypt(m_fileBuffer, [], (err, success) => {
                  console.log("Error", err);
                   var data = Buffer.from(success);
                   string_data = data.toString();
                    // console.log("Success", string_data);
                     )};
                     
)};`
